python/tank/platform/engine.py

```diff
@@ -1367,8 +1517,56 @@ def __load_apps(self):
                                    "validate the configuration loaded from '%s' for app %s. "
                                    "The app will not be loaded." % (self.__env.disk_location, app_instance_name))
                 continue
+
+            # If we're told to reuse existing app instances, check for it and
+            # continue if it's already there. This is most likely a context
+            # change that's in progress, which means we only want to load apps
+            # that aren't already up and running.
+            install_path = descriptor.get_path()
+            if reuse_existing_apps and install_path in self.__application_pool:
+                # If we were given an "old" context that's being switched away
+                # from, we can run the post change method and do a bit of
+                # reinitialization of certain portions of the app.
+                if old_context is not None:
+                    app = self.__application_pool[install_path]
+
+                    try:
+                        # Update the app's internal context pointer.
+                        app.context = self.context
+
+                        # Update the app settings.
+                        app.settings = app_settings
+
+                        # Set the instance name.
+                        app.instance_name = app_instance_name
```

Comment:
We are keying app instances based on the descriptor path.

What if there is 

```
loader1:
   location: {type: app_store, version: 1.2.3 name: tk-multi-loader2}

loader2:
   location: {type: app_store, version: 1.2.3 name: tk-multi-loader2}
```

Seems like loader2 would override loader1 and vice versa?

A unit test for this would be good I think. And in general, unit tests around the restart would be good to assert expected behavior.